### DISCUSSION TITLE
Road to 1.0

### DISCUSSION THREAD

**Discussion by Harris-Miller** [ramda/ramda#3397] [General]
# Preface

Recent discussions in an [breaking MR](https://github.com/ramda/ramda/pull/2938#issuecomment-1521946369) led me to take a look at the state of `ramda` and has led me to want to start a discussion on what the "Road to 1.0" should look like

Out of the last four releases, while 0.26.0, 0.27.0, and 0.28.0 did have breaking changes, they were only for the removal of functions that were deprecated in the previous release. 0.29.0 however, had 2 signature changes. While the rules of semver do dictate that for a [major release zero breaking changes are acceptable for minor releases](https://semver.org/#spec-item-4), continuing to do `0.x` releases for a 9-year-old package at [~10 million downloads a week](https://www.npmjs.com/package/ramda) is becoming problematic. The default carrot ranges for `npm i` will install the latest version, regardless of the current semver range specified, which would lead to unexpected API changes. Tools like depend-a-bot and renovate-bot (to my knowledge) don't treat `0.x` minors as breaking without custom configuration.

So I started to think about what should be done to get to 1.0 and would like to begin a discussion about it

### Disclaimer

I, @Harris-Miller, have only recently begun contributing to `ramda`, particularly for typescript development (see [ramda/types](https://github.com/ramda/types/)). Recent paradigm shifts in front-end architecture, coding style, and state management tools that promote/require immutability, coupled with arrow functions, object/array spread and restructuring, and js bundlers' tree-shaking ability have all led a shift towards FP style javascript and away from classes ever since the release of ES6. I have pushed my peers to adopt FP style javascript and use `ramda` over `lodash`. I've been able to demonstrate how `ramda` truly shines as "A practical functional library for JavaScript". The 2 largest pushback that I've gotten from my peers is the current lack of library support (long times between releases, very old/stale PRs, etc) and poor typescript support. The improvement of typescript support is already a task I have taken on, and so I have a vested interest in the success of `ramda`. And I believe the time is right to invest in that success

# Suggestions

_WIP. Keeping this simple for now with just bullet point suggestions. Goal for right now is to gather feedback. Will update as replies come in_

## Repo and CI/CD

* Updated `CONTRIBUTING` with strict rules concerning PRs
  * PR title naming conventions (eg, `eslint` semantic-release style)
* CI/CD with gitflow (or alternative)
  * `develop` branch that will already release on npm `next` tag, `main` branch on `latest` tag
  * Automate builds and releases across npm, deno, CDNs, etc
* Update all `package.json` dependencies
  * Configure Dependabot for automated maintenance
* Audit and Update the Gitlab Wiki
  * Particularly function naming conventions
* Formalize anything else?

## Code, Issues, and PRs

* What specific PRs would be _required_ before `1.0`
* Utilize Github Milestones more to manage feature release plans
  * Maybe Github projects?
* There are open issues (not including release issues) that date back to 2015
  * Close stale, label others
* Same for PRs
* Built-in typings (`.d.ts` files)
  * Integrate [ramda/types](https://github.com/ramda/types) as previously [proposed](https://github.com/ramda/ramda/discussions/3329)
  * Deprecate `@types/ramda` and `types-ramda`
* Introduce new utility functions
 * eg `noop`, `debounce`/`throttle`
  * Motivation: I don't want to keep using `lodash` for these things

## Documentation

* Type normalization
  * `String` is generally used to represent object keys, other times it is `Idx = String | Int | Symbol`, or `Idx = String | Int`
  * `{k: v}` vs `{s: a}`
* Parameter normalization
  * eg `list` vs `xs` vs `iterable`, `key` vs `prop` vs `param`, `val` vs `x`
* Formal definitions in docs for ADTs
  * `Functor`, `Filterable`, `Applicative`, `Traversable`
  * Add types for functions that support fantasy-land
  * Just link to [fantasy-land](https://github.com/fantasyland/fantasy-land#fantasy-land-specification)?
  * What about non-ADT abstract types like `TypeRep`?
* Removal of unsupported libraries in REPL examples, eg `Ramda-Fantasy` and `Maybe`
  * Alternative: revive and modernize?
  *  Type only alternatives? eg `type Maybe<T> = T | undefined`
  
## What else?

**Comment by kedashoe** [comment]
I reckon I'm on board with most of that, but do we need any of it for 1.0? For me it kinda feels like 1.0 should be ramda as we know it. Then we can get to things like integrated types, es6, releasing to other js platforms, etc.

**Reply by Harris-Miller** [reply]
Sorry, I meant to have replied to this 2 weeks ago. My company did a re-org and all my attention has been to that.

I'm perfectly fine moving to `1.0` now as well. You're right that most of what I stated on that list can easily be done after release. The one exception would be that we should settle https://github.com/ramda/ramda/pull/2938

**Reply by kedashoe** [reply]
I think we're all in agreement? whitelizard summed it up well

> propEq is not really a useful function, as @crshmk has pointed out multiple times above. It could stay in the library, with arg order changed as by this MR, but the library should have the more useful function valAtProp (or whatever would be the best name). Also change the other inconsistent functions (their names!) according to @semmel.

That is why I closed https://github.com/ramda/ramda/pull/3392

We have 10 or so PRs merged since the last release. I could put that out at 0.29.1.

Then we can add the functions from semmel's table, the "properly" named versions of the existing functions. Mark the existing functions as deprecated. Fixes and such can go in as well. Release 0.30.0

Wait a month or two. Fixes etc.

Release 0.31.0

Wait a month or two. Fixes etc.

Release 1.0.0 (2024-01-01?? :stuck_out_tongue: )

**Reply by crshmk** [reply]
> 
> That is why I closed #3392
> 

I thought most people wanted to just revert and keep the existing propEq api. Most seemed to think changing signatures introduced many problems to solve trivial vanity issues. I've used this api for years, and I never saw an issue with it. (Very impressed with it, actually.) I don't understand why [3392](https://github.com/ramda/ramda/pull/3392) was closed.

**Reply by kedashoe** [reply]
I was fine with it as well. But having merged the change, it makes sense to me to keep what we have for now as it fits the more common function naming convention, and update other functions as needed with a smoother transition plan (codemod, deprecation warning, etc). I imagine we'll be adding eqProp in the future so we'll have the preferred order back for this specific function.

**Comment by bunglegrind** [comment]
I'm just a humble observer but most of the work to decide what to put inside 1.0 was done one year (or two?) ago by @customcommander. If I don't make mistakes, issues tagged with "1.0 discussion" were postponed after the 1.0 release. I couldn't find the discussion, but there was some sort of roadmap to 1.0 which was in short: let's release it asap.

**Comment by Harris-Miller** [comment]
Reviving this discussion...

@CrossEye @kedashoe 
CC: @adispring @customcommander @semmel
(please feel free to @ anyone else as a reply, it's honestly hard to tell who is considered "core team", and I'm picking a select few here based on who I see most in replies to Issues/PRs)

I never found the time to follow up on the line items I wrote originally, but I have been hard at work maintaining [ramda/types](https://github.com/ramda/types/). And the recent `0.30.0` release drove me to want to revisit this

In addition to my original post, this discussion [On becoming Ramda 1.0](https://github.com/ramda/ramda/discussions/3344) from a year before this one, the [1.0 discussion labeled issues](https://github.com/ramda/ramda/issues?q=is%3Aopen+is%3Aissue+label%3A%221.0+discussion%22), this [recent comment about getting to 1.0](https://github.com/ramda/ramda/pull/3448#issuecomment-2041764302), I've got a laundry list of small items in my head that I need to get written down here to discuss, eg how to [address fixing releasing to deno](https://github.com/ramda/ramda/issues/3457), integrating [ramda/types](https://github.com/ramda/types/) directly into [ramda/ramda](https://github.com/ramda/ramda), and a faster release cycle to help address higher priority things like [Fix: cross-version placeholder support](https://github.com/ramda/ramda/pull/3463) released fast (seriously that issue is hugely breaking under the right circumstances), just to name a few.

**How would the core team feel about developing a formal Roadmap?**

Utilizing [Github Projects](https://github.com/ramda/ramda/projects), or an alternative, and possibly doing a once-a-month/quarterly Zoom meeting of Owners/Core-Contributors to discuss, as well as act as a steering commit of sorts. (I've been thinking of also building a survey to solicit feedback from ramda users to help kick that off). This would be especially important for issues like continuing to support the `R.__` placeholder (I know there was a whole discussion around this headed by @adispring but for the life of me I cannot find it), as well as realigning on conventions for function names and argument order as [was kicked off by this discussion](https://github.com/ramda/ramda/pull/2938#issuecomment-1521946369)

I would like to bring up the personal concern I have over the continued adoption of `ramda` for typescript projects. This is why I took on the [ramda/types](https://github.com/ramda/types/) myself. Typescript changed the javascript ecosystem in more ways than one, and one of those ways is abandoning the convenient function overloading for discrete type-specific behaviors. (I have a draft write-up for this which I'll post separately in the near future). I feel they also tie into the main concerns that drove the creation of alternative libraries such as [rambda](https://github.com/selfrefactor/rambda), [remeda](https://github.com/remeda/remeda), and [ts-belt](https://github.com/mobily/ts-belt). (You'd think that people would just contribute to `ramda` instead ¯\\\_(ツ)\_\/¯)

Let me know how everyone feels about this idea. Any and all feedback welcome. How would the core team feel about developing a formal Roadmap?

**Comment by Harris-Miller** [comment]
Another issue raised asking for this as well: https://github.com/ramda/ramda/issues/3468

**Comment by kedashoe** [comment]
I personally do not feel like I have time currently to commit to a formal roadmap. I am doing my best to keep the lights on and help move PRs along. The work you are doing on the typescript side is fantastic.

I would be fine committing to any sort of release schedule, I try to maintain the upgrade guide as we go and cutting a release is easy and has gone smoothly recently.

**Comment by Harris-Miller** [comment]
> I personally do not feel like I have time currently to commit to a formal roadmap. I am doing my best to keep the lights on and help move PRs along.

Yeah just dedicating time to `types-ramda` for me is tough. I don't know what I was thinking trying to start up all this on the cusp of summer, the worst time of the year for me in terms of extra-time

I just know that for ~10mil weekly npm downloads, we gotta make some improvements, especially around breaking changes
